File: alexandria/ui/src/io/intino/alexandria/ui/displays/components/Task.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.ui.displays.events.Event;
import io.intino.alexandria.ui.displays.events.ExecuteListener;
import io.intino.alexandria.ui.displays.notifiers.TaskNotifier;

public class Task<DN extends TaskNotifier, B extends Box> extends AbstractTask<DN, B> {
    private String title;
    private String icon;
    private ExecuteListener executeListener;

    public Task(B box) {
        super(box);
    }

    public Task<DN, B> title(String title) {
        this.title = title;
        return this;
    }

    public Task<DN, B> icon(String icon) {
        this.icon = icon;
        return this;
    }

    public void onExecute(ExecuteListener listener) {
        this.executeListener = listener;
    }

    public void execute() {
        if (this.executeListener == null) return;
        this.executeListener.accept(new Event(this));
    }
}

File: framework/ui-elements/src/displays/components/Task.js
import React from "react";
import { withStyles } from '@material-ui/core/styles';
import AbstractTask from "../../../gen/displays/components/AbstractTask";
import TaskNotifier from "../../../gen/displays/notifiers/TaskNotifier";
import TaskRequester from "../../../gen/displays/requesters/TaskRequester";
import Operation from "./Operation"
import { withSnackbar } from 'notistack';
import DisplayFactory from "alexandria-ui-elements/src/displays/DisplayFactory";

class Task extends AbstractTask {
	constructor(props) {
		super(props);
		this.notifier = new TaskNotifier(this);
		this.requester = new TaskRequester(this);
	};
}

export default withStyles(Operation.Styles, { withTheme: true })(withSnackbar(Task));
DisplayFactory.register("Task", withStyles(Operation.Styles, { withTheme: true })(withSnackbar(Task)));